container.go

```diff
@@ -209,25 +209,48 @@ func showContainerCheckpoint(checkpointDirectory, input string) error {
 	return nil
 }
 
-func dirSize(path string) (size int64, err error) {
-	err = filepath.Walk(path, func(_ string, info os.FileInfo, err error) error {
-		if err != nil {
-			return err
+// getCheckpointAndDiffSizes retrieves the size of the checkpoint directory and
```

Comment:
I think I understand your motivation. You loop once over the list of tar files and calculate both sizes.

I am not totally convinced this is the right way from an interface point of view. It feels like two things are returned from one function which do not belong together. At the same time it makes sense this way instead of looping twice over the tar archive.

The only idea I have is, is to have a function which does the same as you do currently, but then have two functions, one which returns the checkpoint size and the other returns the rootfs size. If a size is requested both sizes are calculated and stored in a variable and can then be access by the corresponding function. This sounds, however, like over-engineering...

@rst0git do you have an idea how to do this?

Reply:
What if we have a function that takes a list of string file patterns and returns the total size of the matching files (using the same logic)? This function could then be used within other functions to obtain the size of the `checkpoint` directory or the size of the rootfs tar file.

Reply:
maybe we could use something like the following:

```Go
type archiveSizes struct {
	checkpointSize    int64
	rootFsDiffTarSize int64
}

func GetArchiveSizes(archiveInput string) (*archiveSizes, error) {
	archiveFile, err := os.Open(archiveInput)
	if err != nil {
		return nil, err
	}
	defer archiveFile.Close()

	// Decompress the archive
	decompressedStream, err := archive.DecompressStream(archiveFile)
	if err != nil {
		return nil, err
	}
	defer decompressedStream.Close()

	// Create a tar reader to read the files from the decompressed archive
	tarReader := tar.NewReader(decompressedStream)

	// Initialize the singleton instance
	var result = &archiveSizes{}

	// Iterate through each file in the archive
	for {
		hdr, err := tarReader.Next()
		if errors.Is(err, io.EOF) {
			// End of archive
			break
		}
		if err != nil {
			break
		}

		if hdr.FileInfo().Mode().IsRegular() {
			if strings.HasPrefix(hdr.Name, metadata.CheckpointDirectory) {
				// Add the file size to the total checkpoint size
				result.checkpointSize += hdr.Size
			} else if hdr.Name == metadata.RootFsDiffTar {
				// Assuming there is only one rootfs diff tar file
				result.rootFsDiffTarSize = hdr.Size
			}
		}
	}

	return result, nil
}
```

Reply:
Let's try what @rst0git suggested.